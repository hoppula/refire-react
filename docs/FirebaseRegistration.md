# FirebaseRegistration

> Higher order component for registering users to your Firebase with e-mail and password

## Options

**fields**

Optional. Defines the included fields, defaults are `email` and `password`

**submit** *(dispatch, state) => Promise*

Optional. Custom submit function. By default submitting the form dispatches `createUser` action.

**validator** *(state) => Boolean*

Optional. Custom validator function.

```js
import { FirebaseRegistration } from 'refire-react'
class RegistrationComponent extends Component {
  render() {
    const {
      submit,
      update,
      validInput,
      error,
      processing,
      completed,
      email,
      password
    } = this.props
  }
}
export default FirebaseRegistration({
  fields: ["email", "password"],
  submit: optionalCustomSubmitFn,
  validator: optionalCustomValidatorFn
})(RegistrationComponent)
```
