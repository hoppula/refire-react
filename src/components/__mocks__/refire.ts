import {Â firebaseActions as actions } from "refire"
export { syncFirebase, firebaseToProps, firebaseReducer } from "refire"
export const firebaseActions = {
  ...actions,
  passwordLogin: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  createUser: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  resetPassword: jest.fn((email) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  write: jest.fn((method, path, value, ownProps) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  clearWriteErrors: jest.fn((path) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
}
