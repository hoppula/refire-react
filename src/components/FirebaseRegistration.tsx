import * as React from "react"
import { Component } from "react"
import * as ReactRedux from "react-redux"
import { connect } from "react-redux"
import { firebaseActions, firebaseToProps } from "refire"

const { clearRegistrationError, createUser } = firebaseActions

// default validator assumes that each field has some value
const defaultValidator = (state: ValidatorState): boolean => {
  return Object.keys(state).every((field) => {
    return !!state[field]
  })
}

// default submit dispatches createUser action
const defaultSubmit = (dispatch: ReactRedux.Dispatch<any>, state: RegistrationState) => {
  return dispatch(
    createUser(state.email, state.password)
  ).catch(() => {})
}

interface ValidatorState {
  [key: string]: any
}

export interface RegistrationOptions {
  fields?: string[],
  submit?: (dispatch: ReactRedux.Dispatch<any>, state: RegistrationState) => void,
  validator?: (state: RegistrationState) => boolean
}

export interface Status {
  errors: {
    createUser: string
  },
  processing: {
    createUser: boolean
  },
  completed: {
    createUser: boolean
  }
}

export interface RegistrationProps {
  readonly _status: Status,
  readonly dispatch?: ReactRedux.Dispatch<any>
}

export interface RegistrationState {
  [name: string]: any
}

export interface ExtraProps {
  submit?: (event: React.FormEvent<HTMLElement>) => void,
  update?: (field: string) => (event: React.FormEvent<HTMLInputElement>) => void,
  validInput?: boolean,
  error?: string,
  processing?: boolean,
  completed?: boolean
}

export default function(
  options: RegistrationOptions = {}
) {

  const {
    fields = ["email", "password"],
    submit = defaultSubmit,
    validator = defaultValidator
  } = options

  return function <P>(
    WrappedComponent: ReactRedux.Component<P>
  ) {

    class FirebaseRegistration extends Component<RegistrationProps & P, RegistrationState> {

      constructor(props: RegistrationProps & P) {
        super(props)
        this.state = fields.reduce((initialState, field) => {
          return { ...initialState, [field]: null }
        }, {})
      }

      update = (field: string) => {
        return (event: React.FormEvent<HTMLInputElement>) => {
          const { errors: { createUser: error } } = this.props._status
          if (error) {
            this.props.dispatch(clearRegistrationError())
          }
          this.setState({ [field]: event.currentTarget.value })
        }
      }

      submit = (event: React.FormEvent<HTMLElement>) => {
        event.preventDefault()
        submit(this.props.dispatch, this.state)
      }

      render() {
        const validInput: boolean = validator(this.state)
        const {
          errors: { createUser: error },
          processing: { createUser: processing },
          completed: { createUser: completed }
        } = this.props._status

        const extraProps = {
          ...this.state,
          submit: this.submit,
          update: this.update,
          validInput,
          error,
          processing,
          completed
        }

        return (
          <WrappedComponent
            { ...this.props }
            { ...extraProps }
          />
        )
      }

    }

    return connect<RegistrationProps, void, P>(
      firebaseToProps(["_status"])
    )(FirebaseRegistration)
  }
}
