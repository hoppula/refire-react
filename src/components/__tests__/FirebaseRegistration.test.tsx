import * as React from "react"
import { Provider } from "react-redux"
import * as renderer from "react-test-renderer"
import { applyMiddleware, compose, createStore } from "redux"
import thunk from "redux-thunk"

import { firebaseActions } from "refire"
// createUser action is mocked in ../__mocks__/refire.ts
const { createUser } = firebaseActions

import FirebaseRegistration from "../FirebaseRegistration"

function createEmptyStore() {
  return compose(
    applyMiddleware(thunk)
  )(createStore)((state, action) => {
    return {
      firebase: {
        initialValuesReceived: true,
        errors: {
          createUser: null
        },
        processing: {
          createUser: false
        },
        completed: {
          createUser: false
        }
      }
    }
  })
}

describe("FirebaseRegistration", () => {
  it("update updates chosen field", () => {

    interface UsernameProps {
      username?: string,
      update?: (field: string) => (e) => void,
    }
    function Username({ username, update }: UsernameProps) {
      const changeUsername = (e) => {
        update("username")(e)
      }
      return (
        <div onClick={ changeUsername }>
          { username }
        </div>
      )
    }

    const UsernameComponent = FirebaseRegistration({
      fields: ["email", "password"],
    })(Username)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <UsernameComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({currentTarget: {value: "test@test.com"}})

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("submit calls dispatch prop", () => {

    interface SubmitProps {
      dispatch?: () => void,
      submit?: (event: any) => void
    }
    function Submit({ submit }: SubmitProps) {
      return (
        <button onClick={ submit }>
          Submit
        </button>
      )
    }

    const SubmitComponent = FirebaseRegistration({})(Submit)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <SubmitComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({ preventDefault() {} })
    expect(createUser.mock.calls.length).toBe(1)
  })

  it("validator functions properly")
  it("errors are generated")
  it("processing prop gets updated")
  it("completed prop gets updated")

})




