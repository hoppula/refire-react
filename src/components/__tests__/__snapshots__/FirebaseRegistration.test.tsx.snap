// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseRegistration submit calls dispatch prop 1`] = `
<button
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`FirebaseRegistration update updates chosen field 1`] = `
<div
  onClick={[Function]}
/>
`;

exports[`FirebaseRegistration update updates chosen field 2`] = `
<div
  onClick={[Function]}
>
  test@test.com
</div>
`;
